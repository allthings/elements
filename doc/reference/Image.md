<!-- 
This is an auto-generated markdown. 
You can change it in "src/Image/Image.tsx" and run build:docs to update this file.
-->
# Image
Images make thing more interesting. They can be used
to display user image content and UI graphics.
If something goes wrong when loading the image, a placeholder will
be shown instead.

```example
<Image
   style={{width: 225, height: 225}}
   size="cover"
   src="https://placeimg.com/225/225/people"
/>
```

```example
<Image
   style={{width: 225, height: 225}}
   src="https://placeimg.com/nothing"
/>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|alt|string|Alternative image to use
|src **(required)**||The URL of the image
|srcFallback|string|The URL of the fallback image
|size|"contain", "cover"|The behaviour behavior of image within the container
|position|"center", "left", "right", "top", "bottom"|The position of image
|htmlElement|string|
|alignH|"none", "start", "center", "end", "space-around", "space-between"|horizontal alignment
|alignV|"none", "start", "center", "end", "stretch"|vertical alignment
|direction|"row", "column", "row-reverse", "column-reverse"|direction
|fill|boolean|Passing true, will make the view fill out available space
|wrap|"wrap", "inherit", "initial", "nowrap", "wrap-reverse"|Defining how children will wrap
|flex|number, "flex", "none", "initial", "nogrow", "grow", "auto", "noshrink"|Flex values, can be 5, 10, 15 ... 100 or 33, 66
