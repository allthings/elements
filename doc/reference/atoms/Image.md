<!-- 
This is an auto-generated markdown. 
You can change it in "src/atoms/Image.jsx" and run build:docs to update this file.
-->
# Image
Images make thing more interesting. They can be used
to display user image content and UI graphics.
If something goes wrong when loading the image, a placeholder will
be shown instead.

```example
<Image
  style={{width: 225, height: 225}}
  size="cover"
  src={'https://placeimg.com/225/225/people'}
/>
```

```example
<Image
style={{width: 225, height: 225}}
src={'https://placeimg.com/nothing'}
/>
```
## Usage
| Name        | Type           | Description  |
| ----------- |:--------------:| ------------:|
|alt|string|Alternative image to use
|onClick|func|Will be called when the image is clicked
|src **(required)**|string|The URL of the image
|srcFallback|string|The URL of the fallback image
|size|enum|The Position behavior of image. If passed, image will be rendered as background image.
