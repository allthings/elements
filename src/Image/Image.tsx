import { css } from 'glamor'
import React, { Component, ImgHTMLAttributes } from 'react'
import { ResourceProviderContext } from '../ResourceProvider'
import View, { IViewProps } from '../View'

interface IImage {
  /** Alternative image to use */
  alt?: string
  /** The URL of the image */
  src: string
  /** The URL of the fallback image */
  srcFallback?: string
  /** The behaviour behavior of image within the container */
  size?: 'contain' | 'cover'
  /** The position of image */
  position?: 'center' | 'left' | 'right' | 'top' | 'bottom'
}

type IImageProps = IImage & IViewProps & ImgHTMLAttributes<HTMLElement>

/**
 * Images make thing more interesting. They can be used
 * to display user image content and UI graphics.
 * If something goes wrong when loading the image, a placeholder will
 * be shown instead.
 *
 * ```example
 * <Image
 *   style={{width: 225, height: 225}}
 *   size="cover"
 *   src="https://placeimg.com/225/225/people"
 * />
 * ```
 *
 * ```example
 * <Image
 *   style={{width: 225, height: 225}}
 *   src="https://placeimg.com/nothing"
 * />
 * ```
 */
export default class ImageElement extends Component<IImageProps, {}> {
  state = {
    useFallback: false,
  }

  componentDidMount() {
    this.loadImage(this.props.src)
  }

  componentDidUpdate(prevProps: IImageProps) {
    if (this.props.src !== prevProps.src) {
      this.setState({ useFallback: false })
      this.loadImage(this.props.src)
    }
  }

  loadImage = (src: string) => {
    const image = new window.Image()
    image.onerror = this.onError
    image.src = src
  }

  getFallbackUrl = (resourcePath?: string) => {
    const baseUrl =
      typeof resourcePath === 'undefined'
        ? 'https://static.allthings.me/app/prod/'
        : resourcePath

    return `${baseUrl}/static/img/default/image.svg`
  }

  onError = () => this.props.srcFallback && this.setState({ useFallback: true })

  render() {
    const { srcFallback, src, position, size, ...props } = this.props

    return (
      <ResourceProviderContext.Consumer>
        {({ resourcePath }) => (
          <View
            {...css({
              backgroundImage: `url(${
                this.state.useFallback
                  ? srcFallback || this.getFallbackUrl(resourcePath)
                  : src
              })`,
              backgroundSize: size,
              backgroundPosition: position,
            })}
            {...props}
          />
        )}
      </ResourceProviderContext.Consumer>
    )
  }
}
