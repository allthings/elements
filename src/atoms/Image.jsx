import PropTypes from 'prop-types'
import React from 'react'
import View from '../atoms/View'
import { css } from 'glamor'

/**
 * Images make thing more interesting. They can be used
 * to display user image content and UI graphics.
 * If something goes wrong when loading the image, a placeholder will
 * be shown instead.
 *
 * ```example
 * <Image
 *   style={{width: 225, height: 225}}
 *   size="cover"
 *   src={'https://placeimg.com/225/225/people'}
 * />
 * ```
 *
 * ```example
 * <Image
 * style={{width: 225, height: 225}}
 * src={'https://placeimg.com/nothing'}
 * />
 * ```
 */
export default class Image extends React.Component {
  static propTypes = {
    /** Alternative image to use */
    alt: PropTypes.string,
    /** Will be called when the image is clicked */
    onClick: PropTypes.func,
    /** The URL of the image */
    src: PropTypes.string.isRequired,
    /** The URL of the fallback image */
    srcFallback: PropTypes.string,
    /** The behaviour behavior of image within the container */
    size: PropTypes.oneOf(['contain', 'cover']),
    /** The position of image */
    position: PropTypes.oneOf(['center', 'left', 'right', 'top', 'bottom']),
  }

  state = {
    useFallback: false,
  }

  static contextTypes = {
    resourcePath: PropTypes.string,
  }

  componentDidMount() {
    this.loadImage(this.props.src)
  }

  componentWillReceiveProps(nextProps) {
    this.setState({ useFallback: false })
    if (nextProps.src !== this.props.src) {
      this.loadImage(nextProps.src)
    }
  }

  loadImage = src => {
    const image = new Image()
    image.onerror = this.onError
    image.src = src
  }

  getFallbackUrl = () => {
    const baseUrl =
      typeof this.context.resourcePath === 'undefined'
        ? 'https://static.allthings.me/app/prod/'
        : this.context.resourcePath

    return `${baseUrl}/static/img/default/image.svg`
  }

  onError = () => this.props.srcFallback && this.setState({ useFallback: true })

  render() {
    const { srcFallback, src, position, size, ...props } = this.props

    const imageUrl = this.state.useFallback
      ? srcFallback || this.getFallbackUrl()
      : src

    return (
      <View
        {...css({
          backgroundImage: `url(${imageUrl})`,
          backgroundSize: size,
          backgroundPosition: position,
        })}
        {...props}
      />
    )
  }
}
