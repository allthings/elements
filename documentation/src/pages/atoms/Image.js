import React from 'react'
import Image from '@allthings/elements/atoms/Image'
import View from '@allthings/elements/atoms/View'
import Example from '../../components/Example'
import Notes from '../../components/Notes'

class ImageStory extends React.Component {
  render() {
    return (
      <View>
        <Notes for={Image} />
        <Example>{`<Image
  style={{width: 225, height: 225}}
  size="cover"
  src={'https://placeimg.com/225/225/people'}
/>`}</Example>
        <h3>Broken Image</h3>
        <Example>{`<Image
  style={{width: 225, height: 225}}
  src={'https://placeimg.com/nothing'}
/>`}</Example>
      </View>
    )
  }
}

export default ImageStory
