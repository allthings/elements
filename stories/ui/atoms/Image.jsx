import React from 'react'
import Image from '../../../src/atoms/Image'
import View from '@allthings/react-view'
import { action } from '@storybook/react'
import Example from '../../Example'
import Notes from '../../Notes'

class ImageStory extends React.Component {
  render() {
    return (
      <View>
        <Notes for={Image} />
        <Example>
          <Image
            src={'https://placeimg.com/225/225/people'}
            onClick={action('primary button')}
          />
        </Example>
        <h3>Broken Image</h3>
        <Example>
          <Image
            src={'https://placeimg.com/nothing'}
            onClick={action('primary button')}
          />
        </Example>
      </View>
    )
  }
}

export default ImageStory
