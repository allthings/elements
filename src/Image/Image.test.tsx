import * as React from 'react'
import renderer from 'react-test-renderer'
import Image from './Image'

test('Image renders correctly', () => {
  const tree = renderer
    .create(<Image src="http://example.com/test.js" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})

test('Image should set position', () => {
  const tree = renderer
    .create(<Image src="http://example.com/test.js" position="center" />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
