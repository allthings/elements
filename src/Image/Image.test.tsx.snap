// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders correctly 1`] = `
.css-0,
[data-css-0] {
  background-image: url(http://example.com/test.js);
}

.css-1,
[data-css-1] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

<div
  data-css-1=""
  data-css-0=""
/>
`;

exports[`Image should set position 1`] = `
.css-0,
[data-css-0] {
  box-sizing: border-box;
  flex: 0 0 auto;
  -webkit-flex: 0 0 auto;
}

.css-1,
[data-css-1] {
  background-image: url(http://example.com/test.js);
  background-position: center;
}

<div
  data-css-0=""
  data-css-1=""
/>
`;
